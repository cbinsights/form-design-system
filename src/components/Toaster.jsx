import React, { useState, useEffect } from 'react';
import ReactDOM from 'react-dom';
import { CSSTransition } from 'react-transition-group';
import PropTypes from 'prop-types';

/**
 * @module Toaster
 * Controller for showing, dismissing and positioning Toast components.
 */
const Toaster = ({ toast }) => {
  const [isToasting, setIsToasting] = useState(Boolean(toast));
  let clonedToast = null;

  const dismissToast = () => {
    setIsToasting(false);
  };

  // pass dismiss function down to toast
  if (toast) {
    clonedToast = React.cloneElement(toast, { dismissToast });
  }

  // when the value of toast prop goes from falsey to truthy, set isToasting to true
  useEffect(() => {
    if (toast && !isToasting) {
      setIsToasting(true);
    }
  }, [Boolean(toast)]);

  // set a timeout for toast to dismiss itself
  // eslint-disable-next-line consistent-return
  useEffect(() => {
    if (toast && toast.props.isAutoDismiss) {
      const timer = setTimeout(dismissToast, toast.props.dismissDelay);

      return function cleanup() {
        clearTimeout(timer);
      };
    }
  });
  const onExited = (toast && toast.props && toast.props.onDismiss) || (() => {});

  return ReactDOM.createPortal(
    <div aria-live="assertive" className="toaster">
      <CSSTransition
        in={isToasting}
        appear
        unmountOnExit
        timeout={380}
        classNames="rtgSlideIn"
        onExited={onExited}
      >
        {clonedToast}
      </CSSTransition>
    </div>,
    // eslint-disable-next-line no-undef
    document.body
  );
};

Toaster.propTypes = {
  /** a `Toast` element */
  toast: PropTypes.element,
};

export default Toaster;
