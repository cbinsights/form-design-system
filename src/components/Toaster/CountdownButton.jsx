import React from 'react';
import PropTypes from 'prop-types';
import { Transition } from 'react-transition-group';
import FDS from 'lib/dictionary/js/styleConstants';
import IconButton from 'components/IconButton';
import DenyIcon from 'lib/icons/react/DenyIcon';

/**
 *
 * @param {Number} circleSize bounding box size of circle in px
 * @param {Number} strokeWidth width of circle stroke
 */
export const getCircleInfo = (circleSize, strokeWidth) => {
  const radius = (circleSize - strokeWidth) / 2;

  return {
    r: radius,
    c: Math.PI * radius * 2,
    centerOffset: circleSize / 2,
  };
};

const CountdownButton = ({ label = 'Close', Icon = DenyIcon, duration, onClick }) => {
  const circleSize = 34;
  const strokeWidth = '2';
  const { r, c, centerOffset } = getCircleInfo(circleSize, strokeWidth);

  const defaultCircleStyle = {
    // strokeDasharray sets up a dashed line in which each dash is equal to
    // the circumference of the circle.
    //
    // We offset the dash by 0 to start to create a full circle
    strokeDasharray: c,
    strokeDashoffset: 0,
    transition: `all ${duration}ms linear`,

    // start the countdown at "midnight" position of circle
    transformOrigin: 'center center',
    transform: 'rotate(-90deg)',
  };

  const transitionStyles = {
    entering: { strokeDashoffset: 0 },

    // offsetting the dash by the exact circle circumference lines up
    // the empty part of the dashed stroke with the circle circumference
    entered: { strokeDashoffset: c },
  };

  return (
    <div
      className="countdownButton"
      style={{
        width: `${circleSize}px`,
        height: `${circleSize}px`,
      }}
    >
      <div className="countdownButton-button">
        <IconButton onClick={onClick} radius="circle" Icon={Icon} label={label} />
      </div>
      <Transition
        in={true}
        appear
        timeout={{
          appear: 1,
          enter: duration,
        }}
        mountOnEnter
      >
        {(state) => (
          <svg
            className="countdownButton-animation"
            width={circleSize}
            height={circleSize}
            viewBox={`0 0 ${circleSize} ${circleSize}`}
          >
            <circle
              cx={centerOffset}
              cy={centerOffset}
              r={r}
              strokeWidth={strokeWidth}
              fill="transparent"
              stroke={FDS.FONT_COLOR_HINT}
              style={{
                ...defaultCircleStyle,
                ...transitionStyles[state],
              }}
            />
          </svg>
        )}
      </Transition>
    </div>
  );
};

CountdownButton.propTypes = {
  /** Duration in ms of countdown animation */
  duration: PropTypes.number.isRequired,

  /** Used to render a FDS Icon (should only be used for FDS Icons, e.g. Icon={DenyIcon}) */
  Icon: PropTypes.func,

  /** adds a label to IconButton */
  label: PropTypes.string,

  /** Click callback for button */
  onClick: PropTypes.func,
};

export default CountdownButton;
