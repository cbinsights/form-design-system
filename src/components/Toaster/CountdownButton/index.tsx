import React from 'react';
import { Transition } from 'react-transition-group';
import FDS from 'dictionary/js/styleConstants';
import IconButton from 'components/IconButton';
import DenyIcon from 'icons/react/DenyIcon';
import Icon from 'components/Icon';
export interface CircleInfo {
  radius: number;
  circumference: number;
  centerOffset: number;
}

/**
 *
 * @param {Number} circleSize bounding box size of circle in px
 * @param {Number} strokeWidth width of circle stroke
 */
export const getCircleInfo = (circleSize: number, strokeWidth: number): CircleInfo => {
  const radius = (circleSize - strokeWidth) / 2;

  return {
    radius,
    circumference: Math.PI * radius * 2,
    centerOffset: circleSize / 2,
  };
};

export interface CountdownButtonProps {
  /** Duration in ms of countdown animation */
  duration: number;

  /** Used to render a FDS Icon (should only be used for FDS Icons, e.g. Icon={DenyIcon}) */
  Icon?: Icon;

  /** adds a label to IconButton */
  label?: string;

  /** Click callback for button */
  onClick?: React.MouseEventHandler<HTMLButtonElement>;
}

const CountdownButton = ({
  label = 'Close',
  Icon = DenyIcon,
  duration,
  onClick,
}: CountdownButtonProps): JSX.Element => {
  const circleSize = 34;
  const strokeWidth = 2;
  const { radius, circumference, centerOffset } = getCircleInfo(circleSize, strokeWidth);

  const defaultCircleStyle = {
    // strokeDasharray sets up a dashed line in which each dash is equal to
    // the circumference of the circle.
    //
    // We offset the dash by 0 to start to create a full circle
    strokeDasharray: circumference,
    strokeDashoffset: 0,
    transition: `all ${duration}ms linear`,

    // start the countdown at "midnight" position of circle
    transformOrigin: 'center center',
    transform: 'rotate(-90deg)',
  };

  const transitionStyles = {
    entering: { strokeDashoffset: 0 },

    // offsetting the dash by the exact circle circumference lines up
    // the empty part of the dashed stroke with the circle circumference
    entered: { strokeDashoffset: circumference },
    exiting: undefined,
    exited: undefined,
    unmounted: undefined,
  };

  return (
    <div
      className="countdownButton"
      style={{
        width: `${circleSize}px`,
        height: `${circleSize}px`,
      }}
    >
      <div className="countdownButton-button">
        <IconButton onClick={onClick} radius="circle" Icon={Icon} label={label} />
      </div>
      <Transition
        in={true}
        appear
        timeout={{
          appear: 1,
          enter: duration,
        }}
        mountOnEnter
      >
        {(state) => (
          <svg
            className="countdownButton-animation"
            width={circleSize}
            height={circleSize}
            viewBox={`0 0 ${circleSize} ${circleSize}`}
          >
            <circle
              cx={centerOffset}
              cy={centerOffset}
              r={radius}
              strokeWidth={strokeWidth}
              fill="transparent"
              stroke={FDS.FONT_COLOR_HINT}
              style={{
                ...defaultCircleStyle,
                ...transitionStyles[state],
              }}
            />
          </svg>
        )}
      </Transition>
    </div>
  );
};

export default CountdownButton;
