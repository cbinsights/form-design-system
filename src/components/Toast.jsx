import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import IconButton from './buttons/IconButton';
import Button from './buttons/Button';
import CountdownButton from './buttons/CountdownButton';
import Flex from './layout/Flex';
import FlexItem from './layout/FlexItem';
import customPropTypes from './util/customPropTypes';
import FDS from '../../lib/dictionary/js/styleConstants';
import CheckIcon from '../../lib/icons/react/CheckIcon';
import InformationIcon from '../../lib/icons/react/InformationIcon';
import DenyIcon from '../../lib/icons/react/DenyIcon';

const Toast = ({
  content,
  type,
  actionLabel,
  progress,
  onAction,
  isAutoDismiss,
  dismissDelay,
  canDismiss,
  dismissToast /* Passed from Toaster */,
}) => {
  const classNames = cx('toast', `toast--${type}`);

  const icon = {
    info: <InformationIcon size="l" color={FDS.COLOR_AQUA} />,
    success: <CheckIcon size="xs" color={FDS.COLOR_WHITE} />,
    warn: (
      <div role="image" className="fontWeight--bold color--white">
        !
      </div>
    ),
    error: (
      <div role="image" className="fontWeight--bold color--white">
        !
      </div>
    ),
  }[type];

  const useContinuousProgressBar = type === 'progress' && !progress;
  const useAmountProgressBar = type === 'progress' && progress;

  const toastBar = (
    <div className="toast-bar">
      {useContinuousProgressBar && (
        <div className="toast-progressBar toast-progressBar--continuous" />
      )}
      {useAmountProgressBar && (
        <div
          className="toast-progressBar toast-progressBar--amount"
          style={{ width: `${progress}%` }}
        />
      )}
    </div>
  );

  const onActionDismiss = (event) => {
    dismissToast();
    if (typeof onAction === 'function') {
      onAction(event);
    }
  };

  return (
    <div
      className={classNames}
      role={type === 'error' ? 'alert' : 'log'}
      aria-relevant="all"
      aria-atomic="true"
    >
      {toastBar}
      <div className="toast-content alignChild--left--center">
        <Flex>
          {icon && (
            <FlexItem shrink>
              <div className="toast-constrainGrowth alignChild--center--center">
                <div className="toast-icon media--s">{icon}</div>
              </div>
            </FlexItem>
          )}
          <FlexItem>
            <div className="toast-constrainGrowth">{content}</div>
          </FlexItem>
          {actionLabel && onAction && (
            <FlexItem shrink>
              <div className="toast-constrainGrowth alignChild--center--center">
                <Button onClick={onActionDismiss} data-test="toast-action" theme="ghost">
                  {actionLabel}
                </Button>
              </div>
            </FlexItem>
          )}
          {canDismiss && type !== 'progress' && (
            <FlexItem shrink>
              <div className="toast-constrainGrowth alignChild--center--center">
                {isAutoDismiss ? (
                  <CountdownButton onClick={dismissToast} duration={dismissDelay} />
                ) : (
                  <IconButton Icon={DenyIcon} onClick={dismissToast} radius="circle" />
                )}
              </div>
            </FlexItem>
          )}
        </Flex>
      </div>
    </div>
  );
};

Toast.defaultProps = {
  dismissDelay: 4000,
  canDismiss: true,
  isAutoDismiss: true,
  onDismiss: () => {},
};

Toast.propTypes = {
  /** JSX Content of Toast */
  content: PropTypes.element.isRequired,

  /** Type of toast */
  type: PropTypes.oneOf(['info', 'warn', 'error', 'success', 'progress']).isRequired,

  /** Label for action button */
  actionLabel: PropTypes.string,

  /** Callback for action button click */
  onAction: PropTypes.func,

  /** Callback invoked on dismissal of toast */
  onDismiss: PropTypes.func,

  /** Time in ms to auto-dismiss toast */
  dismissDelay: PropTypes.number,

  /** Should this toast auto-dismiss itself? */
  isAutoDismiss: PropTypes.bool,

  /** Is this toast user-dismissable? */
  canDismiss: PropTypes.bool,

  /** Number 1-100 declaring % progress */
  progress: customPropTypes.range(1, 100),

  /** Passed by Toaster: function to dismiss the toast */
  dismissToast: PropTypes.func,
};

export default Toast;
