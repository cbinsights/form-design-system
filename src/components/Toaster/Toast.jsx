import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import FDS from 'lib/dictionary/js/styleConstants';
import CheckIcon from 'lib/icons/react/CheckIcon';
import InformationIcon from 'lib/icons/react/InformationIcon';
import DenyIcon from 'lib/icons/react/DenyIcon';
import Flex from 'components/Flex';
import FlexItem from 'components/FlexItem';
import customPropTypes from 'components/util/customPropTypes';
import Button from 'components/Button';
import IconButton from 'components/IconButton';

import CountdownButton from './CountdownButton';

export const TYPES = ['info', 'warn', 'error', 'success', 'progress'];

const Toast = ({
  isAutoDismiss = true,
  dismissDelay = 4000,
  canDismiss = true,
  dismissOnAction = true,
  type = 'info',
  content,
  actionLabel,
  progress,
  onAction,
  dismissToast /* Passed from Toaster */,
}) => {
  const classNames = cx('toast', `toast--${type}`);

  const icon = {
    info: <InformationIcon size="l" color={FDS.COLOR_AQUA} />,
    success: <CheckIcon size="xs" color={FDS.COLOR_WHITE} />,
    warn: (
      <div role="image" className="fontWeight--bold color--white">
        !
      </div>
    ),
    error: (
      <div role="image" className="fontWeight--bold color--white">
        !
      </div>
    ),
  }[type];

  const useContinuousProgressBar = type === 'progress' && !progress;
  const useAmountProgressBar = type === 'progress' && progress;

  const toastBar = (
    <div className="toast-bar">
      {useContinuousProgressBar && (
        <div className="toast-progressBar toast-progressBar--continuous" />
      )}
      {useAmountProgressBar && (
        <div
          className="toast-progressBar toast-progressBar--amount"
          style={{ width: `${progress}%` }}
        />
      )}
    </div>
  );

  const onActionDismiss = (event) => {
    if (dismissOnAction) dismissToast();
    if (onAction) onAction(event);
  };

  return (
    <div
      className={classNames}
      role={type === 'error' ? 'alert' : 'log'}
      aria-relevant="all"
      aria-atomic="true"
    >
      {toastBar}
      <div className="toast-content alignChild--left--center">
        <Flex>
          {icon && (
            <FlexItem shrink>
              <div className="toast-constrainGrowth alignChild--center--center">
                <div className="toast-icon media--s">{icon}</div>
              </div>
            </FlexItem>
          )}
          <FlexItem>
            <div className="toast-constrainGrowth">{content}</div>
          </FlexItem>
          {actionLabel && (onAction || (dismissToast && dismissOnAction)) && (
            <FlexItem shrink>
              <div className="toast-constrainGrowth alignChild--center--center">
                <Button
                  onClick={onActionDismiss}
                  data-test="toast-action"
                  theme="ghost"
                  label={actionLabel}
                />
              </div>
            </FlexItem>
          )}
          {canDismiss && type !== 'progress' && (
            <FlexItem shrink>
              <div className="toast-constrainGrowth alignChild--center--center">
                {isAutoDismiss ? (
                  <CountdownButton
                    onClick={dismissToast}
                    duration={dismissDelay}
                    label="Close"
                  />
                ) : (
                  <IconButton
                    Icon={DenyIcon}
                    onClick={dismissToast}
                    radius="circle"
                    label="Close"
                  />
                )}
              </div>
            </FlexItem>
          )}
        </Flex>
      </div>
    </div>
  );
};

Toast.propTypes = {
  /** JSX Content of Toast */
  content: PropTypes.node.isRequired,

  /** Type of toast */
  type: PropTypes.oneOf(TYPES),

  /** Label for action button */
  actionLabel: PropTypes.string,

  /** Callback for action button click */
  onAction: PropTypes.func,

  /** Time in ms to auto-dismiss toast */
  dismissDelay: PropTypes.number,

  /** Should this toast auto-dismiss itself? */
  isAutoDismiss: PropTypes.bool,

  /* Controls if toast gets dismiss when the ActionButton is clicked */
  dismissOnAction: PropTypes.bool,

  /** Is this toast user-dismissable? */
  canDismiss: PropTypes.bool,

  /** Number 1-100 declaring % progress */
  progress: customPropTypes.range(1, 100),

  /** Passed by Toaster: function to dismiss the toast */
  dismissToast: PropTypes.func,
};

export default Toast;
