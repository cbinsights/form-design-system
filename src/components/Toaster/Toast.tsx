import React from 'react';
import cc from 'classcat';
import FDS from 'dictionary/js/styleConstants';
import CheckIcon from 'icons/react/CheckIcon';
import InformationIcon from 'icons/react/InformationIcon';
import DenyIcon from 'icons/react/DenyIcon';
import Flex from 'components/Flex';
import FlexItem from 'components/FlexItem';
import Button from 'components/Button';
import IconButton from 'components/IconButton';

import CountdownButton from './CountdownButton';

type ToastType = 'info' | 'warn' | 'error' | 'success' | 'progress';

export interface ToastProps {
  /** JSX Content of Toast */
  content: React.ReactNode;

  /** Type of toast */
  type?: ToastType;

  /** Label for action button */
  actionLabel?: string;

  /** Callback for action button click */
  onAction?: React.MouseEventHandler<HTMLButtonElement>;

  /** Time in ms to auto-dismiss toast */
  dismissDelay?: number;

  /** Should this toast auto-dismiss itself? */
  isAutoDismiss?: boolean;

  /* Controls if toast gets dismiss when the ActionButton is clicked */
  dismissOnAction?: boolean;

  /** Is this toast user-dismissable? */
  canDismiss?: boolean;

  /** Number 1-100 declaring % progress */
  progress?: number;

  /** Passed by Toaster: function to dismiss the toast */
  dismissToast?(...args: unknown[]): unknown;
}

const Toast = ({
  isAutoDismiss = true,
  dismissDelay = 4000,
  canDismiss = true,
  dismissOnAction = true,
  type = 'info',
  content,
  actionLabel,
  progress,
  onAction,
  dismissToast /* Passed from Toaster */,
}: ToastProps): JSX.Element => {
  const classNames = cc(['toast', `toast--${type}`]);

  const icon = {
    info: <InformationIcon size="l" color={FDS.COLOR_WHITE} />,
    success: <CheckIcon size="xs" color={FDS.COLOR_WHITE} />,
    warn: (
      <div role="image" className="fontWeight--bold color--white">
        !
      </div>
    ),
    error: (
      <div role="image" className="fontWeight--bold color--white">
        !
      </div>
    ),
    progress: undefined,
  }[type];

  const accentColor = {
    info: FDS.COLOR_BLUE,
    warn: FDS.COLOR_ORANGE,
    error: FDS.COLOR_RED,
    success: FDS.COLOR_GREEN,
    progress: FDS.COLOR_AQUA,
  }[type];

  const iconStyle = { backgroundColor: accentColor };

  const onActionDismiss: React.MouseEventHandler<HTMLButtonElement> = (event) => {
    if (dismissOnAction && dismissToast) dismissToast();
    if (onAction) onAction(event);
  };

  return (
    <div
      className={classNames}
      role={type === 'error' ? 'alert' : 'log'}
      aria-relevant="all"
      aria-atomic="true"
    >
      <div className="toast-bar">
        <svg
          xmlns="http://www.w3.org/2000/svg"
          className={cc([
            'toast-svg',
            { 'toast-svg--animate': type === 'progress' && !progress },
          ])}
        >
          <line
            x1="0%"
            y1="50%"
            x2={type === 'progress' && progress ? `${progress}%` : '100%'}
            y2="50%"
            stroke={accentColor}
            strokeWidth={FDS.SPACE_S}
            strokeDasharray={type === 'progress' && !progress ? '200 400' : ''}
          />
        </svg>
      </div>
      <div className="toast-content alignChild--left--center">
        <Flex>
          {icon && (
            <FlexItem shrink>
              <div className="toast-constrainGrowth alignChild--center--center">
                <div className="toast-icon media--s" style={iconStyle}>
                  {icon}
                </div>
              </div>
            </FlexItem>
          )}
          <FlexItem>
            <div className="toast-constrainGrowth">{content}</div>
          </FlexItem>
          {actionLabel && (onAction || (dismissToast && dismissOnAction)) && (
            <FlexItem shrink>
              <div className="toast-constrainGrowth alignChild--center--center">
                <Button
                  onClick={onActionDismiss}
                  data-test="toast-action"
                  theme="ghost"
                  label={actionLabel}
                />
              </div>
            </FlexItem>
          )}
          {canDismiss && type !== 'progress' && (
            <FlexItem shrink>
              <div className="toast-constrainGrowth alignChild--center--center">
                {isAutoDismiss ? (
                  <CountdownButton
                    onClick={dismissToast}
                    duration={dismissDelay}
                    label="Close"
                  />
                ) : (
                  <IconButton
                    Icon={DenyIcon}
                    onClick={dismissToast}
                    radius="circle"
                    label="Close"
                  />
                )}
              </div>
            </FlexItem>
          )}
        </Flex>
      </div>
    </div>
  );
};

export default Toast;
