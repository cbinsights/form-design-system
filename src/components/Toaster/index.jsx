import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import ReactDOM from 'react-dom';
import { CSSTransition, TransitionGroup } from 'react-transition-group';
import Toast from './Toast';

export const Toaster = ({
  toastProps = {},
  dismissDelay = 4000,
  isAutoDismiss = true,
  isOpen = false,
  id,
  onDismiss,
}) => {
  useEffect(() => {
    // We need to explicitly check of canDismiss is NOT true, as it has a default
    // parameter of true that this component does not know about
    if (
      isAutoDismiss &&
      toastProps.type !== 'progress' &&
      toastProps.canDismiss !== false
    ) {
      const timer = setTimeout(() => onDismiss(), dismissDelay);

      return function cleanup() {
        clearTimeout(timer);
      };
    }

    return undefined;
  }, [id, isOpen, isAutoDismiss, toastProps.type, toastProps.canDismiss]);

  return (
    <div aria-live="assertive">
      <TransitionGroup>
        {isOpen && (
          <CSSTransition
            key={id}
            appear
            unmountOnExit
            timeout={380}
            classNames="rtgSlideIn"
          >
            <div className="toaster">
              <Toast
                dismissDelay={dismissDelay}
                dismissToast={onDismiss}
                isAutoDismiss={isAutoDismiss}
                content={toastProps.content}
                type={toastProps.type}
                progress={toastProps.progress}
                canDismiss={toastProps.canDismiss}
                actionLabel={toastProps.actionLabel}
                onAction={toastProps.onAction}
                dismissOnAction={toastProps.dismissOnAction}
              />
            </div>
          </CSSTransition>
        )}
      </TransitionGroup>
    </div>
  );
};

const ToasterWrapper = (props) =>
  ReactDOM.createPortal(
    <Toaster {...props} />,
    // eslint-disable-next-line no-undef
    document.body
  );

Toaster.propTypes = {
  /**  A UUID for the toast. Each new toast should have a new UUID. */
  id: PropTypes.string,
  /** Specifies whether Toast is open or not */
  isOpen: PropTypes.bool,
  /** Accepts all props that `Toast` accepts. Refer to Toast component for full documentation */
  toastProps: PropTypes.shape(Toast.propTypes),
  /** Should this toast auto-dismiss itself? */
  isAutoDismiss: PropTypes.bool,
  /** Time in ms to auto-dismiss toast */
  dismissDelay: PropTypes.number,
  /** Callback that is fired when toast is dismissed */
  onDismiss: PropTypes.func,
};

export default ToasterWrapper;
