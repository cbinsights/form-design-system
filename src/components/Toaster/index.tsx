import React, { useEffect } from 'react';
import ReactDOM from 'react-dom';
import { CSSTransition, TransitionGroup } from 'react-transition-group';
import Toast, { ToastProps } from './Toast';

const dismissDelay = 4000;

export interface ToasterProps {
  /**  A UUID for the toast. Each new toast should have a new UUID. */
  id?: string;

  /** Specifies whether Toast is open or not */
  isOpen?: boolean;

  /** Should this toast auto-dismiss itself? */
  isAutoDismiss?: boolean;

  onDismiss?: (...args: unknown[]) => void;

  /** Controls the display for your specific toast. */
  toastInstance?: ToastProps;
}

export const Toaster = ({
  isAutoDismiss = true,
  isOpen = false,
  onDismiss,
  toastInstance = {
    content: null,
  },
  id,
}: ToasterProps): JSX.Element => {
  useEffect(() => {
    // We need to explicitly check of canDismiss is NOT true, as it has a default
    // parameter of true that this component does not know about
    if (
      isAutoDismiss &&
      toastInstance.type !== 'progress' &&
      toastInstance.canDismiss !== false &&
      onDismiss
    ) {
      const timer = setTimeout(() => onDismiss(), dismissDelay);

      return function cleanup() {
        clearTimeout(timer);
      };
    }

    return undefined;
  }, [
    id,
    isOpen,
    isAutoDismiss,
    toastInstance.type,
    toastInstance.canDismiss,
    onDismiss,
  ]);

  return (
    <div aria-live="assertive">
      <TransitionGroup>
        {isOpen && (
          <CSSTransition
            key={id}
            appear
            unmountOnExit
            timeout={380}
            classNames="rtgSlideIn"
          >
            <div className="toaster">
              <Toast
                dismissDelay={dismissDelay}
                isAutoDismiss={isAutoDismiss}
                dismissToast={onDismiss}
                content={toastInstance.content}
                type={toastInstance.type}
                progress={toastInstance.progress}
                canDismiss={toastInstance.canDismiss}
                actionLabel={toastInstance.actionLabel}
                onAction={toastInstance.onAction}
                dismissOnAction={toastInstance.dismissOnAction}
              />
            </div>
          </CSSTransition>
        )}
      </TransitionGroup>
    </div>
  );
};

const ToasterWrapper = (props: ToasterProps): React.ReactNode =>
  ReactDOM.createPortal(
    <Toaster {...props} />,
    // eslint-disable-next-line no-undef
    document.body
  );

export default ToasterWrapper;
