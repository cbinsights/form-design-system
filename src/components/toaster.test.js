import React from 'react';
import { mount, shallow } from 'enzyme';

import Toast from './Toast';
import Toaster from './Toaster';

const renderToaster = (useShallow, props) => {
  const toast = <Toast type="info" content={<p>test</p>} />;
  let result;
  if (useShallow) {
    result = shallow(<Toaster toast={toast} {...props} />);
  } else {
    result = mount(<Toaster toast={toast} {...props} />);
  }
  return result;
}


describe('Toaster component', () => {

  it('matches snapshot', () => {
    const component = renderToaster(true);
    expect(component).toMatchSnapshot();
  });

  it('passes dismiss func to toast', () => {
    const component = renderToaster();
    const {dismissToast} = component.find(Toast).props();
    expect(typeof dismissToast).toBe('function');
  })

});
