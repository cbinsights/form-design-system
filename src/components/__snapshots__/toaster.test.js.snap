// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toaster component matches snapshot 1`] = `
<Portal
  containerInfo={<body />}
>
  <div
    aria-live="assertive"
    className="toaster"
  >
    <CSSTransition
      appear={true}
      classNames="rtgSlideIn"
      in={true}
      onExited={[Function]}
      timeout={380}
      unmountOnExit={true}
    >
      <Toast
        canDismiss={true}
        content={
          <p>
            test
          </p>
        }
        dismissDelay={4000}
        dismissToast={[Function]}
        isAutoDismiss={true}
        onDismiss={[Function]}
        type="info"
      />
    </CSSTransition>
  </div>
</Portal>
`;
